SI-8512 Infer a type for f"$args"

som-snytt opened the pull request:
The f-interpolator gets a type param that better be Any to avoid
unfortunate widenings.

Hey, it worked!

@densh 's comments on the ticket are longer than the fix and the test.

Review by @densh

som-snytt committed:
SI-8512 Infer a type for f"$args"

The f-interpolator gets a type param that better be Any to avoid
unfortunate widenings.

Hey, it worked!

Unfortunately, when `Any` is inferred, `-Xlint:infer-any` takes notice.

This is probably a greater problem for the f-interpolator than for
quasiquotes, which are a more specialized tool.

xeno-by commented on src/library/scala/StringContext.scala:
```diff
@@ -163,7 +163,7 @@ case class StringContext(parts: String*) {
   // The implementation is hardwired to `scala.tools.reflect.MacroImplementations.macro_StringInterpolation_f`
   // Using the mechanism implemented in `scala.tools.reflect.FastTrack`
-  def f(args: Any*): String = macro ???
+  def f[A >: Any](args: A*): String = macro ???
```
Which kind of widenings does this prevent? Could you elaborate a bit more?

som-snytt replied:
In the unit test, f"${3.14}%.2f rounds to ${3}%d", `undetParam inferred: type A as Double` and 

```
performing macro expansion scala.StringContext.apply("", "%f rounds to ", "%d").f[Double](3.14, 3.0)
```

Which I take to mean the 3 becomes 3.0 before the macro sees it.  I haven't checked yet for anything that enables undoing that conversion (like an attachment or history of some kind).  Maybe a lazy Sunday afternoon is a good time to read some code.

densh replied:
That's quite surprising. It turns out that quasiquotes have the same issue here:

```
scala> q"${1} + ${1.0}"
res1: reflect.runtime.universe.Tree = 1.0.$plus(1.0)
```

Please add this bound to quasiquote interpolator definition too https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/api/Quasiquotes.scala#L16

som-snytt replied:
No problemo.

som-snytt commented:
Jenkins couldn't find 2.10.2.  Who uses 2.10.2 anymore?  Moreover, Jenkins was unable to send an alert email to typesafe to let them know.  Doesn't Jenkins know that typesafe takes August off?  Since Roman times, August has been known as the least typesafe month of the year.

gourlaysama commented:
PLS REBUILD/pr-scala@450e78a

gourlaysama commented:
Maven central seems to be fixed, let's see if that works...

scala-jenkins commented:
(kitty-note-to-self: ignore 51042996)
:cat: Roger! Rebuilding pr-scala for 450e78a1. :rotating_light:

som-snytt committed:
SI-8512 Infer Any for the q

Avoid the widening bug for q.  This resolution also suffers
from the inference of Any, which can trigger a warning and
an anxiety attack.  But that's still better than doing the
wrong thing.

Right?

gkossakowski commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 51482879)
:cat: Roger! Rebuilding pr-scala for 450e78a1, 4a6d47c2. :rotating_light:

som-snytt commented:
Rebased, conflict at added test functions. Now that's something it could figure out with minimal smarts.

densh commented:
LGTM

gkossakowski commented:
Thanks!
